sentinelhub/download/session.py

```diff
@@ -126,3 +134,125 @@ def _fetch_token(self, request: DownloadRequest) -> JsonDict:
+        access the memory before it even exists."""
+        super().start()
+        self._is_memory_shared_event.wait()
+
+    def run(self) -> None:
+        """A running thread is running an infinite loop of sharing a token and waiting for token to expire. The loop
+        ends only when the thread is stopped."""
+        self._stop_event.clear()
+
+        while not self._stop_event.is_set():
+            token = self.session.token
+            self._share_token(token)
+
+            sleep_until_refresh_time = token["expires_at"] - time.time() - self._refresh_time
+            if sleep_until_refresh_time > 0:
+                self._stop_event.wait(timeout=sleep_until_refresh_time)
+
+    def _share_token(self, token: JsonDict) -> None:
+        """A token is encoded into bytes and written into a shared memory block.
+
+        Note that the `SharedMemory` object allocates extra `self._EXTRA_MEMORY_BYTES` bytes of memory because the
+        length of encoded token can vary a bit.
+        """
+        encoded_token = json.dumps(token).encode()
+
+        if self._is_memory_shared_event.is_set():
+            memory = SharedMemory(name=self.memory_name)
+        else:
+            memory = SharedMemory(
+                create=True,
+                size=len(encoded_token) + self._EXTRA_MEMORY_BYTES,
+                name=self.memory_name,
+            )
+            self._is_memory_shared_event.set()
+
+        try:
+            memory.buf[:] = encoded_token + _NULL_MEMORY_VALUE * (memory.size - len(encoded_token))
+        finally:
+            memory.close()
+
+    def stop(self) -> None:
+        """The method stops the thread that would otherwise run indefinitely.
+
+        After the stop even is set it is important to wait for `run` method to finish and only afterward unlink shared
+        memory.
+        """
+        self._stop_event.set()
+        self._wait_for_tstate_lock()  # type: ignore[attr-defined]
+
+        if self._is_memory_shared_event.is_set():
+            memory = SharedMemory(name=self.memory_name)
+            memory.unlink()
+            self._is_memory_shared_event.clear()
+            memory.close()
+
+
+def collect_shared_session(memory_name: str = _DEFAULT_SESSION_MEMORY_NAME) -> SentinelHubSession:
+    """This utility function is meant to be used in combination with `SessionSharingThread`. It retrieves an
+    authentication token from the shared memory and returns it in an `SentinelHubSession` object.
+
+    :param memory_name: A unique name of the requested shared memory block from where to read the session. It should
+        match the one used in `SessionSharingThread`.
+    :return: An instance of `SentinelHubSession` that contains the shared token but is not self-refreshing.
+    """
+    memory = SharedMemory(name=memory_name)
+    try:
+        encoded_token = memory.buf.tobytes().rstrip(_NULL_MEMORY_VALUE)
```

Comment:
it cannot happen that this would remove too much right?

Reply:
Nope because the last valid byte will always be `b"}"`.

Reply:
alright